os/se/ameba/security_ameba_wrapper.c

```diff
@@ -0,0 +1,1933 @@
+	0x53, 0xac, 0xe2, 0xee, 0x32, 0x39, 0xe3, 0x93, 0xe4, 0xa, 0x9a, 0xaf, 0xc, 0x12, 0x53, 0xdd,
+	0x3c, 0x16, 0xd5, 0x63, 0xdf, 0xd6, 0x9f, 0xf3, 0xe3, 0x55, 0xe5, 0xf9, 0x8, 0x47, 0xfb, 0xe3,
+	0xd9, 0x1f, 0x30, 0xa, 0xa7, 0xc4, 0xd9, 0xa9, 0x5c, 0x95, 0x5b, 0xd5, 0xdf, 0xf8, 0x32, 0xce,
+	0x24, 0x7d, 0x6, 0xcb, 0x52, 0xe4, 0xf, 0x1c, 0xde, 0x73, 0x63, 0x79, 0x40, 0x5d, 0x75, 0x95,
+	0xa3, 0x70, 0xec, 0xa9, 0xbf, 0xe0, 0xdf, 0x68, 0xbc, 0x5a, 0x39, 0x95, 0x8f, 0x9, 0xc7, 0x9c,
+	0xa0, 0xc3, 0xd3, 0x2f, 0x6f, 0x5b, 0xa1, 0x67, 0xa7, 0x8e, 0x43, 0xf0, 0xae, 0x42, 0x67, 0x43,
+	0xcc, 0xda, 0x67, 0xd1, 0x86, 0xa6, 0xd3, 0xfe, 0xf0, 0xf0, 0x38, 0xb8, 0xc9, 0xd7, 0x32, 0xc4,
+	0xcc, 0x87, 0x9, 0x74, 0xff, 0xe7, 0x39, 0xcb, 0x69, 0x6a, 0x9f, 0xbf, 0xac, 0x6a, 0x3d, 0x85,
+	0xc4, 0xc6, 0xc2, 0xa2, 0x56, 0xa4, 0x2, 0x0, 0x47, 0xa5, 0x48, 0xf2, 0xcd, 0x37, 0x90, 0x6e,
+	0x5b, 0x60, 0xde, 0xd0, 0x29, 0xd4, 0xe8, 0xbe, 0x62, 0x88, 0x1a, 0xed, 0xbe, 0xba, 0x6b, 0xb5,
+	0xd9, 0x84, 0x11, 0x59, 0x59, 0xb4, 0x1d, 0x20, 0x2c, 0xa8, 0x25, 0xfb, 0x6c, 0x36, 0x4d, 0x7d,
+	0xf0, 0xaf, 0xa, 0x1c, 0xb6, 0xba, 0xde, 0x55, 0x71, 0x4, 0xb4, 0x3d, 0x2b, 0x20, 0xd9, 0x4e,
+	0xea, 0x98, 0x6f, 0xbc, 0x45, 0x80, 0x5d, 0xb6, 0x41, 0xbf, 0xed, 0x6d, 0x1c, 0x4f, 0x62, 0xb0,
+	0x93, 0x10, 0x1b, 0xa3, 0x54, 0x67, 0x52, 0xd4, 0x21, 0xaa, 0xdf, 0xc8, 0xa8, 0xaa, 0x9e, 0xf9,
+	0xd1, 0x21, 0x2b, 0x6d, 0x73, 0xec, 0xb5, 0x6b, 0x3e, 0x42, 0x53, 0xfe, 0xa4, 0x29, 0x49, 0x14,
+	0x6d, 0x32, 0xa7, 0x14, 0xc5, 0xa1, 0xfe, 0xc, 0x8f, 0xb3, 0x64, 0x85, 0xd, 0x4c, 0x9a, 0xbb,
+	0xab, 0xdf, 0x67, 0xe9, 0x47, 0xd1, 0x4c, 0xe1, 0xd, 0xc0, 0x3c, 0xeb, 0x7d, 0xb5, 0x10, 0x6,
+	0xd8, 0x8e, 0xc2, 0xe0, 0x31, 0xc7, 0xb3, 0x92, 0x28, 0xdf, 0xf3, 0x96, 0xb0, 0xb, 0x3c, 0xf5,
+	0xe8, 0x5d, 0x2b, 0xaf, 0xdf, 0xe9, 0x2f, 0xf8, 0x55, 0xd2, 0xf4, 0x99, 0x2, 0x3c, 0x7e, 0x30,
+	0x65, 0xaf, 0xc5, 0x68, 0x76, 0x33, 0x6a, 0xc1, 0xf2, 0x9a, 0x9f, 0xc5, 0xa3, 0x9c, 0xae, 0xe8,
+	0xc0, 0x4c, 0x7, 0xd0, 0xc5, 0x6a, 0x39, 0xb6, 0xa8, 0x1b, 0x38, 0xfd, 0xaa, 0xe1, 0x10, 0x7e,
+	0x2a, 0x82, 0x9c, 0x38, 0x16, 0xa5, 0x4a, 0xa3, 0x58, 0x57, 0xe1, 0xf, 0x7c, 0xba, 0x5b, 0x96,
+	0x6d, 0x6a, 0x1e, 0x99, 0x9c, 0x3e, 0xfd, 0x1, 0xf6, 0x40, 0xfc, 0xd4, 0x1d, 0x89, 0x74, 0xe3,
+	0xd0, 0xcb, 0x67, 0x1b, 0xb2, 0x72, 0xb2, 0xd9, 0x1, 0x35, 0xd7, 0x17, 0x12, 0x6f, 0xfa, 0x1
+};
+
+static const unsigned char RSA_4096_E[4] = {
+	0x0, 0x1, 0x0, 0x1
+};
+
+typedef struct key_struc {
+	hal_key_type key_type;
+
+	/* AES Key must be 4Byte Aligned, due to hardware crypto */
+	unsigned char aes_key[32] __attribute__((aligned(4)));
+	uint32_t aes_key_len;
+
+	rtl_rsa_context rsa_ctx;
+	rtl_ecdh_context ecdh_ctx;
+	rtl_ecdsa_context ecdsa_ctx;
+	rtl_dhm_context dhm_ctx;
+
+	unsigned char hmac_key[32];
+
+	unsigned char *cert_data;
+	uint32_t cert_len;
+} key_struc;
+
+static key_struc key_storage[KEY_STORAGE_INDEX_MAX] = { 0 };
+static uint32_t init_stat = HAL_NOT_INITIALIZED;
```

Comment:
why is key storage not in SE? the key storage should be protected.

Reply:
All key is store in SE in struct array format, and initially are empty. Init and gen key have to be call to generate the Key. 
In future SE will be putting in trust zone, all SE function and data will be protected.